README.md
# Beefcake

A pure-Ruby Google Protocol Buffers library. It's all about being Buf; ProtoBuf.

## Installation

```shell
gem install beefcake
```

## Usage

```ruby
require 'beefcake'

class Variety
  include Beefcake::Message

  # Required
  required :x, :int32, 1
  required :y, :int32, 2

  # Optional
  optional :tag, :string, 3

  # Repeated
  repeated :ary,  :fixed64, 4
  repeated :pary, :fixed64, 5, :packed => true

  # Enums - Simply use a Module (NOTE: defaults are optional)
  module Foonum
    A = 1
    B = 2
  end

  # As per the spec, defaults are only set at the end
  # of decoding a message, not on object creation.
  optional :foo, Foonum, 6, :default => Foonum::B
end

# You can create a new message with hash arguments:
x = Variety.new(:x => 1, :y => 2)

# You can set fields individually using accessor methods:
x = Variety.new
x.x = 1
x.y = 2

# And you can access fields using Hash syntax:
x[:x] # => 1
x[:y] = 4
x # => <Variety x: 1, y: 4>
```

### Encoding

Any object responding to `<<` can accept encoding

```ruby
# see code example above for the definition of Variety
x = Variety.new(:x => 1, :y => 2)

# For example, you can encode into a String:
s = ""
x.encode(s)
s # => "\b\x01\x10\x02)\0"

# If you don't encode into anything, a new Beefcake::Buffer will be returned:
x.encode # => #<Beefcake::Buffer:0x007fbfe1867ab0 @buf="\b\x01\x10\x02)\0">

# And that buffer can be converted to a String:
x.encode.to_s # => "\b\x01\x10\x02)\0"
```

### Decoding

```ruby
# see code example above for the definition of Variety
x = Variety.new(:x => 1, :y => 2)

# You can decode from a Beefcake::Buffer
encoded = x.encode
Variety.decode(encoded) # => <Variety x: 1, y: 2, pary: [], foo: B(2)>

# Decoding from a String works the same way:
Variety.decode(encoded.to_s) # => <Variety x: 1, y: 2, pary: [], foo: B(2)>

# You can update a Beefcake::Message instance with new data too:
new_data = Variety.new(x: 12345, y: 2).encode
Variety.decoded(new_data, x)
x # => <Variety x: 12345, y: 2, pary: [], foo: B(2)>
```

### Generate code from `.proto` file

```shell
protoc --beefcake_out output/path -I path/to/proto/files/dir path/to/file.proto
```

You can set the `BEEFCAKE_NAMESPACE` variable to generate the classes under a
desired namespace. (i.e. App::Foo::Bar)

## About

Ruby deserves and needs first-class ProtoBuf support. Other libs didn't feel
very "Ruby" to me and were hard to parse.

This library was built with EventMachine in mind.  Not just blocking-IO.

Source: https://github.com/protobuf-ruby/beefcake

### Support Features

  * Optional fields
  * Required fields
  * Repeated fields
  * Packed Repeated Fields
  * Varint fields
  * 32-bit fields
  * 64-bit fields
  * Length-delimited fields
  * Embedded Messages
  * Unknown fields are ignored (as per spec)
  * Enums
  * Defaults (i.e. `optional :foo, :string, :default => "bar"`)
  * Varint-encoded length-delimited message streams

### Future

  * Imports
  * Use package in generation
  * Groups (would be nice for accessing older protos)

### Further Reading

http://code.google.com/apis/protocolbuffers/docs/encoding.html

## Testing

    rake test

Beefcake conducts continuous integration on [Travis CI](http://travis-ci.org).
The current build status for HEAD is [![Build Status](https://travis-ci.org/protobuf-ruby/beefcake.png?branch=master)](https://travis-ci.org/protobuf-ruby/beefcake).

All pull requests automatically trigger a build request.  Please ensure that
tests succeed.

Currently Beefcake is tested and working on:

  * Ruby 1.9.3
  * Ruby 2.0.0
  * Ruby 2.1.0
  * Ruby 2.1.1
  * Ruby 2.1.2
  * JRuby in 1.9 mode

## Thank You

  * Keith Rarick (kr) for help with encoding/decoding.
  * Aman Gupta (tmm1) for help with cross VM support and performance enhancements.


RELEASE_NOTES.md
# Beefcake Release Notes

# 1.2.0 - 2016-06-21

Release 1.2.0 includes one bug fix.

* Clone strings before `force_encoding`. [#70](https://github.com/protobuf-ruby/beefcake/pull/70)

# 1.1.0 - 2015-05-26

Release 1.1.0 includes improvements and bug fixes.

* Correct message-hash conversion behavior, fixed by Peter "kybu" Vrabel.
* Remove trailing whitespace in generator output without package name.
* Allow a Hash instead of an Array<Hash> when assigning a repeated field that's
  only repeated once.
* Ensure object identities are persisted through message initialization,
  reported by Peter Neubauer.

# 1.0.0 - 2014-09-05

Release 1.0.0 includes changes and improvements.

* Version numbering now properly semantic.
* Ruby 1.8 is no longer supported.
* Field number re-use raises a `DuplicateFieldNumber` error.
* Checking to see if a type is encodable is much faster.
* Fields named `fields` are now supported.
* String read and decoding are benchmarked during testing.
* `string` fields now decode with a `UTF-8` encoding.

# 0.5.0 - 2013-12-20

Release 0.5.0 corrects a few behaviors.

* Drastically revised README, written by Tobias "grobie" Schmidt
* Output fewer newlines in generated files, fixed by Tobias "grobie" Schmidt
* Don't crash when attempting to reencode frozen strings,
  found thanks to Kyle "Aphyr" Kingsbury
* Return `nil` instead of raising a generic Ruby error when trying to
  decode a zero-length buffer, fixed by Tobias "grobie" Schmidt

# 0.4.0 - 2013-10-10

Release 0.4.0 is the first with new maintainers.

* Modernize tests
* Add Travis CI monitoring
* Support varint-encoded length-delimited buffers
* Support generation with recursive definitions
* Support Ruby 2.0
* Support encoded buffers
* Support false but non-nil values
* Support top-level enums, added by Kim Altintop:
  https://github.com/protobuf-ruby/beefcake/pull/23


bench/simple.rb
$:.unshift File.expand_path('../../lib', __FILE__)
require 'beefcake'

class MyMessage
  include Beefcake::Message
  required :number, :int32,  1
  required :chars,  :string, 2
  required :raw,    :bytes,  3
  required :bool,   :bool,   4
  required :float,  :float,  5
end

ITERS = 100_000

case ARGV[0]
when 'pprof'
  # profile message creation/encoding/decoding w/ perftools.rb
  # works on 1.8 and 1.9
  #   ruby bench/simple.rb pprof
  #   open bench/beefcake.prof.gif

  ENV['CPUPROFILE_FREQUENCY'] = '4000'
  require 'rubygems'
  require 'perftools'
  PerfTools::CpuProfiler.start(File.expand_path("../beefcake.prof", __FILE__)) do
    ITERS.times do
      str = MyMessage.new(
        :number => 12345,
        :chars  => 'hello',
        :raw    => 'world',
        :bool   => true,
        :float  => 1.2345
      ).encode
      MyMessage.decode(str)
    end
  end
  Dir.chdir(File.dirname(__FILE__)) do
    `pprof.rb beefcake.prof --gif > beefcake.prof.gif`
  end

when 'ruby-prof'
  # profile message creation/encoding/decoding w/ ruby-prof
  # works on 1.8 and 1.9
  #   ruby bench/simple.rb ruby-prof
  #   open bench/beefcake.prof.html

  require 'ruby-prof'
  result = RubyProf.profile do
    ITERS.times do
      str = MyMessage.new(
        :number => 12345,
        :chars  => 'hello',
        :raw    => 'world',
        :bool   => true,
        :float  => 1.2345
      ).encode
      MyMessage.decode(str)
    end
  end

  filename = File.expand_path('beefcake.prof.html', File.dirname(__FILE__))
  File.open(filename, 'w') do |file|
    RubyProf::GraphHtmlPrinter.new(result).print(file)
  end

else
  # benchmark message creation/encoding/decoding
  #   rvm install 1.8.7 1.9.2 jruby rbx
  #   rvm 1.8.7,1.9.2,jruby,rbx ruby bench/simple.rb

  require 'benchmark'

  Benchmark.bmbm do |x|
    x.report 'object creation' do
      ITERS.times do
        Object.new
      end
    end
    x.report 'message creation' do
      ITERS.times do
        MyMessage.new(
          :number => 12345,
          :chars  => 'hello',
          :raw    => 'world',
          :bool   => true,
          :float  => 1.2345
        )
      end
    end
    x.report 'message encoding' do
      m = MyMessage.new(
        :number => 12345,
        :chars  => 'hello',
        :raw    => 'world',
        :bool   => true,
        :float  => 1.2345
      )
      ITERS.times do
        m.encode
      end
    end
    x.report 'message decoding' do
      str = MyMessage.new(
        :number => 12345,
        :chars  => 'hello',
        :raw    => 'world',
        :bool   => true,
        :float  => 1.2345
      ).encode.to_s
      ITERS.times do
        MyMessage.decode(str.dup)
      end
    end
  end

end
